### DISCUSSION TITLE
GSoC '22 contributors WELCOME!

### DISCUSSION THREAD

**Discussion by ClaudiaComito** [helmholtz-analytics/heat#919] [General]
Thanks for your interest! 

------
**UPDATE Mar 30th 2022**

A summary of Heat resources:
- Our [GSoC22 contributor guidance](https://docs.google.com/document/d/1Wd3cSazkvefAyuFHMb_wxGBXXh00ETqXVvQTELBLVwQ)
- Our [docs](https://heat.readthedocs.io/en/latest/introduction.html) including [Tutorials](https://heat.readthedocs.io/en/latest/tutorials.html)
- [Jupyter Notebook](https://github.com/helmholtz-analytics/heat/blob/main/scripts/tutorial.ipynb) illustrating the basics
- More details on [contributing to Heat](https://github.com/helmholtz-analytics/heat/blob/main/contributing.md) 
- [Parallel programming / MPI / mpi4py](https://github.com/helmholtz-analytics/heat/discussions/934) resources
- we held an AMA on March 29th. You can access it on our [Mattermost](https://mattermost.hzdr.de/signup_user_complete/?id=3sixwk9okpbzpjyfrhen5jpqfo) (sign up with your GitHub credentials, join the GSoC22 channel)

------

Here's our [List of Ideas](https://docs.google.com/document/d/1OKUf_n_whN0M20c__jj8R5jyGABPYYxmdBfyzto4OZk/), and our primary mentors:

1. Python [array-API standard](https://data-apis.org/array-api/latest/) compatibility @mtar 
2. Memory-distributed sparse operations @ClaudiaComito 
3. Memory-distributed 2D convolution @ClaudiaComito 
4. Memory-distributed Singular Value Decomposition @coquelin77 
5. Data-Parallel Neural Network Training with [DASO](https://doi.org/10.1186/s40537-021-00556-1) @coquelin77 
6. [mpi4torch](https://github.com/helmholtz-analytics/mpi4torch): An automatic-differentiable MPI wrapper for pytorch @d1saster (Philipp Knechtges)
7. Support HIP/ROCm @mtar 
8. Interactive benchmarking analysis @ClaudiaComito and @bhagemeier 

We're looking forward to hearing from you! 

**EDIT**: feel free to sign up to our [Mattermost](https://mattermost.hzdr.de/signup_user_complete/?id=3sixwk9okpbzpjyfrhen5jpqfo) (you can sign up with your GitHub credentials) and join the **GSoC22** channel.

**Comment by Mystic-Slice** [comment]
Hi!
Hope you are all doing well!
I am Ashwath. I am a sophomore undergraduate student majoring in Computer Science Engineering. I came across your project and found it very interesting. I am particularly interested in "Heat for Neuroscience: Memory-distributed sparse operations" (Project Idea 2) and "Data-Parallel Neural Network Training with DASO" (Project Idea  5). I would love to know more about the projects. Should I ask my questions here or would it be better if I talked with the respective mentors through the mail?
Also, if you have any questions for me, please feel free to ask!

**Reply by ClaudiaComito** [reply]
@Mystic-Slice Ashwath thanks a lot for your interest. Please ask away!

**Reply by Mystic-Slice** [reply]
Hey! Sorry for the late reply. 
I carefully studied the [HeAT research paper](https://arxiv.org/abs/2007.13552). It was amazing to see how much more efficient HeAT is compared to other libraries. I have some questions to better understand HeAT and what you would expect from me during the course of the project.

**General questions related to HeAT:**

~~**1.** How do I go about understanding the HeAT codebase? Shall I start working on any issues that I could pick off?~~

~~**2.** The best way to work on HeAT?  like, how to set up the development environment and such?~~
Right now, the best way I can think of is to create a Conda environment and work on the library there. I can install HeAT while I am working on it without it affecting any global modules "outside".

**3.** There are many papers cited in the reference. I don't know if I can go through all of them.  Do you think any of those cited papers would help me understand the project better? If so, can you point me towards those select few?

**Reply by Mystic-Slice** [reply]
**Questions about the DASO project:**

**1.** Is there a list of sorts for the features that you would expect this project to implement?

**2.** Just to be certain, NCCL is [Nvidia's Communications library](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/index.html), right?

~~**3.** The skill requirements state that I would have to know parallel/distributed computing. I have a basic understanding of parallel programming. I learned it as a part of a Java course that I took in college. I have good experience working with asynchronous programming, which is not exactly parallel programming, but close! I have worked on apps built on Qt(C++) and Electron.js(JavaScript), Android(Java) frameworks where I have employed async programming to make the apps more responsive. I have used a message-passing model as well for communication. I would like to learn more about parallel programming. Are there any resources that you would suggest? I am currently going through [this course](https://www.coursera.org/learn/scala-parallel-programming) to know more.~~ **[ANSWERED]**

**Reply by Mystic-Slice** [reply]
**Questions about Memory-distributed sparse operations:**

**1.** As far as I understand, I have to implement HeAT's version of the [Pytorch sparse](https://pytorch.org/docs/stable/sparse.html). Am I right in assuming that? If so, will I be more focussed on implementing [these](https://pytorch.org/docs/stable/sparse.html#supported-linear-algebra-operations) functionalities along with the sparse data structure?

**2.** What is meant by the memory-distributed linear operator mentioned in the project description? 
I tried searching about it but could not find anything. Maybe, I did find it and did not know that I was looking at it.

**Reply by Mystic-Slice** [reply]
Going through the project ideas again. I liked **"Heat for SKA: memory-distributed 2D convolution"** also. 
_"An implementation of one-dimensional distributed convolution for Heat is in progress"_
What is the status of this? Is the code available anywhere for me to view? If it is still in progress, maybe in the developers' fork or smth?

**Reply by Mystic-Slice** [reply]
Sorry for flooding this discussion with questions. I ask these questions only because I am very interested. I hope you don't find it a nuisance.

**Reply by ClaudiaComito** [reply]
> Going through the project ideas again. I liked **"Heat for SKA: memory-distributed 2D convolution"** also. _"An implementation of one-dimensional distributed convolution for Heat is in progress"_ What is the status of this? Is the code available anywhere for me to view? If it is still in progress, maybe in the developers' fork or smth?

see #939

**Comment by samiali12** [comment]
Hi, my Self Sami Ali, currently pursuing BS in Software Engineering at COMSATS UNIVERSITY OF ISLAMABAD, WAH CAMPUS. I came across your project and found it very interesting. I am particularly interested in Data-Parallel Neural Network Training with DASO (Project Idea 5), Memory-distributed sparse operations (Project Idea 2). I would love to know more about the projects. Should I ask my questions here or would it be better if I talked with the respective mentors through the mail?

**Comment by tikka01** [comment]
Hi! I am Chaitika, an undergraduate student majoring in Mathematics. I'm interested in Project idea 01- Python array-API standard compatibility and I was hoping that I could connect with someone with whom I can discuss the project in depth.

**Comment by ClaudiaComito** [comment]
Thank you so much @Mystic-Slice @samiali12 @tikka01 for your questions! Several general topics are arising, I'll collect them in separate threads. 

- [Parallel programming resources](https://github.com/helmholtz-analytics/heat/discussions/934) 

More later!

**Reply by tikka01** [reply]
Hi Claudia , I was hoping that maybe there is someway that I can get in touch with my primary mentor @mtar for discussing Idea -1.

**Comment by Mystic-Slice** [comment]
Hi @ClaudiaComito. I asked some questions about the project ideas a few days ago. I think you might have missed these questions since the thread is collapsed.
Memory distributed sparse operations: https://github.com/helmholtz-analytics/heat/discussions/919#discussioncomment-2392560
Memory distributed 2-D Convolution: https://github.com/helmholtz-analytics/heat/discussions/919#discussioncomment-2392566
I am more interested in these two projects. Would love to discuss the ideas more.

Also, I am not able to access [Mattermost](https://mattermost-haf.fz-juelich.de/signup_email). Any idea why that might be?

**Reply by ClaudiaComito** [reply]
Good catch, that is an old Mattermost link. Try [this one](https://mattermost.hzdr.de/signup_user_complete/?id=3sixwk9okpbzpjyfrhen5jpqfo), you can sign up with your GitHub credential, and join the GSoC22 channel.

I'll get to your other questions tomorrow.

**Comment by sahilvora10** [comment]
Hi!


I am Sahil Vora. I am currently working as a Software Developer, and I have completed my undergraduate studies majoring in Computer and Communication Engineering. 
I came across your project and organization and found it very interesting. 
I am particularly interested in Heat general: Python array-API standard compatibility (Project 1) and would love to know more about the projects. Should I ask my questions here, or is there any other channel I could join? 
Also, if you have any questions for me, please feel free to ask!

**Comment by ClaudiaComito** [comment]
Ping, original post updated

**Comment by RATED-R-SUNDRAM** [comment]
Hello Team,  I am shivam sundram 3rd year undergraduate at IIT Mandi, I have 3 years (including one year of professional experience in python). I have been an active contributor in open source and would like to contribute in this community for the this year's GSOC. The idea I am intrested in and would like to contribute is the (PROJECT IDEA 7) . Please help me in getting started and refering me to resources required for this project.

**Reply by RATED-R-SUNDRAM** [reply]
Hey @mtar please provide some input on the project 7 so I can finalize the draft making.

**Reply by mtar** [reply]
Hi @RATED-R-SUNDRAM, have you looked at the links to PyTorch and our devices.py file in the List of Ideas? As stated there, we want to use PyTorch's HIP interface for AMD GPU support and set up a testing environment eventually.

Questions regarding idea 7 can now be posed in Discussion #963.

**Comment by SnehaK253** [comment]
Hello, I am Sneha first year undergraduate student from India. I am interested in project idea 4 (heat for engineering: memory- distributed SVD) and project 1(Heat general: python array-API standard compatibility). I want to know in depth of project idea 4. Kindly provide me resources related to this.

**Comment by anirudha-mirmira** [comment]
Hi, I am Anirudha Mirmira, a final year theoretical physics PhD student in JNCASR, India. I work in Condensed Matter physics and predominantly write Fortran programs implementing different methods to solve non-equilibrium quantum systems. I have used MPI for parallelization in my code and have used convolutions (1d) via FFT in my calculations. I have working knowledge of Python and have basic familiarity with the pytorch framework, though am very interested in learning more about the same. I also have some familiarity with OpenACC and OpenMP. This is also the first time I have come across the open-source community and am very enthused to contribute. 

I am interested in the sparse matrix and the 2D convolution projects and am very excited to have an opportunity to work on the same. Coming to the time commitment, since I am finalising my thesis, I have quite a hectic schedule and will only be able to dedicate a couple of hours every day in the evenings (IST- UTC+5:30). I'm looking forward to contributing both as a part of GSoC and otherwise.

**Comment by BohrCh** [comment]
Hello, everyone! 
I am Bo Chen, a 2nd-year graduate student at the University of Science & Technology Beijing. My graduation research project is "Molecular dynamics optimization on heterogeneous processors" I use the MPI and the OpenMP in my project. I found this organization by searching "high-performance computing" I was an intern at Baidu Inc. last year. I was responsible for benchmarking the performance of cuDNN on the A100. So I am interested in the "Interactive benchmarking analysis via Ginko Performance Explorer Description"(PROJECT IEDA 8) and "Mpi4torch: An automatic-differentiable MPI wrapper for PyTorch"(PROJECT IEDA 6). Is it too late to apply these projects? @d1saster @ClaudiaComito  @bhagemeier  @ClaudiaComito

**Reply by bhagemeier** [reply]
Dear @BohrCh , sorry for the late reply. It should never be too later for handing in something before the deadline. If you prepared anything, you'd have to hand it in today. The contributor application deadline is at 18:00 UTC.

**Comment by aditya-shrivastavv** [comment]
helmholtz-analytics | is there any updates regarding GSoC 2023

**Reply by ClaudiaComito** [reply]
Hi @aditya-shrivastavv thanks for your interest! 

We will certainly apply as GSoC Mentoring Organization again. We will know on Feb 22nd whether we are in. Feel free to check out our [Issues](https://github.com/helmholtz-analytics/heat/issues) in the meantime, we'll keep posting new issues related to potential GSoC projects.
